### ISSUE TITLE
MQTT in Node: Error stopping node: Close timed out

### ISSUE THREAD

**Issue by hybuild-project** [node-red/node-red#2934]
### What are the steps to reproduce?
Import the flow in: https://cookbook.nodered.org/mqtt/connect-to-broker, start any MQTT broker (e.g. Mosquitto) and deploy
EDIT: The error is not evident with Mosquitto, as initially reported, but e.g. with https://github.com/martin-doyle/node-red-contrib-aedes.

### What happens?
Every time the flow is deployed, the MQTT in node gets stuck for a while, then output: "Error stopping node: Close timed out"

### What do you expect to happen?
The MQTT in node is immediately closed without errors and the flow is deployed without error and time wasting.

### Please tell us about your environment:

- [ ] Node-RED version: 1.3.2
- [ ] Node.js version: 14.6.1
- [ ] npm version: 
- [ ] Platform/OS: Win 10
- [ ] Browser: Chrome

**Comment by knolleary** [comment]
Hi,

I'm not able to reproduce this issue, so there may be something more specific with your environment that is contributing to it.

I assume you have the broker running locally rather than using a remote broker?

**Comment by hardillb** [comment]
Are using mosquito 2.0? And is it on the same machine as node-red?

**Comment by hybuild-project** [comment]
Thanks Nick, thanks Ben, today it seems that everything is working well. Yesterday I tried many times, restarting many times, too. Anyway, I use Mosquitto 1_6_12a on the same machine. I will close the issue now. If it will happen again I will check and keep you in the loop. Thanks again.

**Comment by hybuild-project** [comment]
P.S. Seems like the problem arises if you have used Aedes node (https://flows.nodered.org/node/node-red-contrib-aedes) in your flows, no matter if enabled, disabled or erased. I will inform the authors of the node about. I leave the comment for who may encounter the same issue.

**Comment by hybuild-project** [comment]
The error is connected to https://github.com/node-red/node-red/commit/04a3c4bb222f1708877ddd98379be801bfc6fc72. It happens only when a full deploy, not incremental.
Full credits for the doublecheck go to https://github.com/martin-doyle who I thanks again.
Waiting for v1.3.3.
Thanks for your excellent work!

**Comment by martin-doyle** [comment]
Hi,

I am the maintainer of [node-red-contrib-aedes](https://github.com/martin-doyle/node-red-contrib-aedes). This issue only happens with a broker as part of Node-RED. An external broker is fine. 
As far as I can see Node-RED closes the MQTT broker node-red-contrib-aedes first. Then the MQTT nodes send a disconnect which can't be answered anymore since the broker is already down.

`20 Apr 20:20:43 - [info] [aedes broker:Standard Broker] Unbinding aedes mqtt server from port: 1885`
`20 Apr 20:20:43 - [info] [mqtt-broker:localhost:1885] Disconnected from broker: mqtt://localhost:1885`
`20 Apr 20:20:58 - [error] [mqtt in:80486cfb.0747d] Error stopping node: Close timed out`

As mentioned before, everything was fine until version 1.2.9. The issue only happens when Node-RED shuts down i.e.  at close and at a full deploy.

Is there anything I can do in order to make the broker close after the MQTT nodes? Or do I have to send a specific signal before close?

No change for version 1.3.3. Any help is welcome.

Martin

**Comment by kuema** [comment]
Could the change https://github.com/node-red/node-red/pull/2880 be relevant this?

**Comment by knolleary** [comment]
Yes, this will be related to the change in order of stopping things.

Previous, the flow/config nodes would be stopped in a somewhat random order - related to the order they were originally added to the workspace.

Now, the flow nodes are stopped first and *then* the config nodes are stopped.

This originally broke the 'on-close' message sending of the MQTT nodes as it was being handled by the close event of the config node - by which time all of the MQTT nodes would have already been stopped and the connection closed. So the sending of the message was moved to get triggered when the last MQTT node is closed.

If the aedes broker is a flow node, it will getting stopped alongside the MQTT nodes. If the aedes broker node was added first, it will get stopped before all of the MQTT nodes - meaning the connection will be dropping in parallel to the MQTT nodes stopping and trying to send the final on-close message.

To be honest, I'm not sure why you'd have an on-close message configured in the MQTT nodes if your broker is running inside Node-RED as well - the only time they are going to disconnect cleanly is when the flow is restarting or Node-RED is stopping, in which case the broker is going to be restarting/stopping as well.

**Comment by martin-doyle** [comment]
As far as I know, sending the on-close message is the standard behaviour of the MQTT config nodes. There is no additonal message configured. So, the problem is now that the MQTT config node expects an answer of a broker which is already down. Right now, I have no idea how to fix this issue on the broker side.

**Comment by knolleary** [comment]
Ah right. I thought this was related to the publishing of the `on-close` message (which is what the referenced commit fixed).

This is just the node timing out when trying to disconnect.. which is indeed odd. I'll investigate.

**Comment by hardillb** [comment]
The message Nick is referring to in the last paragraph is the one configured in the "Messages" tab for the MQTT-in/out's config node that includes the broker connection details.

These are the Birth/Close/LWT messages. It doesn't make much sense to have any (OK, may be the Birth message) of these messages if the broker is in the same process as the client as they will only fire when the broker is very likely to be in the process of restarting as well.

**Comment by emc02** [comment]
having the same issue on RaspberryPi with:
Node-RED version: v1.3.5
node.js version: v12.22.1
npm version: 6.14.12
Platform/OS: RaspiOS
Browser: Chrome and Firefox

**Comment by hardillb** [comment]
@emc02 Just to be clear, what broker are you using?

**Comment by emc02** [comment]
I am using aedes broker and connecting to zigbee2mqtt
It worked fine till I upgraded yesterday to latest version of nodered and broker.

**Comment by daniel-schel** [comment]
I have this problem too, with an external broker on a separate VM. I use for this:
- vernemq v1.11.0
- Node-Red v1.3.5

The MQTT broker node is configured with the MQTT protocol version 3.1.1 and there are no messages configured on connection, before disconnecting or unexpected disconnection.

**Comment by martin-doyle** [comment]
The issue still exists with Node-RED version 2. Is there any fix available?

**Comment by RGNagel** [comment]
Same problem here. Any update? Thanks!

**Comment by denonbw** [comment]
Hello Together
after updating from 2.2.0 to 2.2.1 I have the same Behavior. 
The installation few weeks old and has only this small flow:

``
[
    {
        "id": "150d20998fc7fe72",
        "type": "tab",
        "label": "Flow 1",
        "disabled": false,
        "info": "",
        "env": []
    },
    {
        "id": "391e5b8123b3fdc0",
        "type": "mqtt in",
        "z": "150d20998fc7fe72",
        "name": "",
        "topic": "stepperin/status/stepperrun",
        "qos": "2",
        "datatype": "auto",
        "broker": "98d6b4dad9e2c400",
        "nl": false,
        "rap": true,
        "rh": 0,
        "inputs": 0,
        "x": 290,
        "y": 60,
        "wires": [
            [
                "a4fc05cefd74869d"
            ]
        ]
    },
    {
        "id": "a4fc05cefd74869d",
        "type": "debug",
        "z": "150d20998fc7fe72",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "x": 550,
        "y": 60,
        "wires": []
    },
    {
        "id": "e7d268159853c70f",
        "type": "mqtt out",
        "z": "150d20998fc7fe72",
        "name": "",
        "topic": "stepperin",
        "qos": "",
        "retain": "",
        "respTopic": "",
        "contentType": "",
        "userProps": "",
        "correl": "",
        "expiry": "",
        "broker": "98d6b4dad9e2c400",
        "x": 520,
        "y": 180,
        "wires": []
    },
    {
        "id": "3848d22e6f02fa97",
        "type": "inject",
        "z": "150d20998fc7fe72",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "{ \"aktion\": \"movestepper\", \"stepps\": \"21968\", \"speed\": \"10000\", \"acceleration\": \"30000\" }",
        "payloadType": "str",
        "x": 250,
        "y": 160,
        "wires": [
            [
                "e7d268159853c70f"
            ]
        ]
    },
    {
        "id": "f84e8543482cf0dd",
        "type": "inject",
        "z": "150d20998fc7fe72",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "{ \"aktion\": \"movestepper\", \"stepps\": \"-21968\", \"speed\": \"10000\", \"acceleration\": \"30000\" }",
        "payloadType": "str",
        "x": 250,
        "y": 220,
        "wires": [
            [
                "e7d268159853c70f"
            ]
        ]
    },
    {
        "id": "98d6b4dad9e2c400",
        "type": "mqtt-broker",
        "name": "localhost",
        "broker": "localhost",
        "port": "1883",
        "clientid": "nodered",
        "autoConnect": true,
        "usetls": false,
        "protocolVersion": "4",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "0",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "sessionExpiry": ""
    }
]
``

The broker runs on the same pi and is mosquitto (2.0.11).
The update was done by script:
[https://nodered.org/docs/getting-started/raspberrypi](https://nodered.org/docs/getting-started/raspberrypi)

I have already completely deleted the flow and recreated by hand. The first 4 or 5 deploys were fast. After that the behavior happened again.

**Comment by hardillb** [comment]
@denonbw This issue is all about using [node-red-contrib-aedes](https://github.com/martin-doyle/node-red-contrib-aedes).

If you have a problem when using an external broker (mosquitto) then it is very unlikely to be related to this as it is caused by the broker (aedes) being shutdown before the mqtt node connected to it tries to send it's last message.

Given your flow doesn't appear to be using birth or close messages again not likely to be related to this issue. I suggest you open a new issue and include all the relevant data (including the Node-RED log output covering the period before and after the deploy)

**Comment by sanyafifa** [comment]
Same problem after update
> 2.2.0 to 2.2.1
 mosquitto version 2.0.13
nodejs v16.14.0

**Comment by hardillb** [comment]
@sanyafifa  can you supply Node-RED logs from before and after a deploy and the broker logs from the same period. Along with the type of deploy (full, changed flows, changed nodes).

**Comment by Steve-Mcl** [comment]
@sanyafifa 

As well as you logs, can you describe you setup (OS, Hardware, direct connection to internet? behind proxy?) 
AND 
provide a list of installed contrib nodes

**Comment by Paul-Reed** [comment]
Same here since updating from v2.2.0 to v2.2.1.

![mqtt](https://user-images.githubusercontent.com/973580/154165966-ab391a91-0b0d-4d24-8ed7-67884d8a1032.JPG)

**Comment by Steve-Mcl** [comment]
@Paul-Reed 

Hi Paul, I see aedes in your screenshot but I don't say it listed in your nodes. 

What gives?

Are you using aedes as a broker in your node-red?

**Comment by Paul-Reed** [comment]
Aedes broker is running on a cloud server, which has not been edited, or even the server accessed during the past few weeks.
The client is running on my local Pi, using the default MQTT 'in node'
The error message occurred immediately after updating my local Pi from  v2.2.0 to v2.2.1 (in the local Pi debug), and every time I restart flows.

The node-RED log shows this problem when my local Pi closes down;

```
15 Feb 23:49:42 - [info] Stopping flows
15 Feb 23:49:42 - [info] serial port /dev/ttyAMA0 closed
15 Feb 23:49:57 - [error] [mqtt in:Aedes test] Error stopping node: Close timed out
15 Feb 23:49:57 - [error] [mqtt in:Dummy] Error stopping node: Close timed out
15 Feb 23:49:57 - [error] [mqtt in:3e51925c4519d148] Error stopping node: Close timed out
15 Feb 23:49:57 - [info] Stopped flows
```

When node_RED restarts, If I search amongst my Cron+ log entries... :mute:
The only Aedes entry I see is;

`15 Feb 23:49:59 - [info] [mqtt-broker:aedes mqtt] Connected to broker: mqtt://140.238.65.67:1883`

**Comment by Paul-Reed** [comment]
In fact, it's not confined to Aedes.
In the  stopping flows log (my last post above) the entry;
`15 Feb 23:49:57 - [error] [mqtt in:Dummy] Error stopping node: Close timed out`
That is a standard MQTT 'in node', which is connected to the remote `cloudmqtt.com` broker on port 15069.

Both Aedes & cloudmqtt have been happily working for a number of months, and these errors have only started after updating to node-RED v2.2.1, and worked fine under v2.2.0.

**Comment by Paul-Reed** [comment]
If you can't replicate, I'm around today, and will gladly help wherever I can.

**Comment by denonbw** [comment]
For me it is not an pressing problem, so I can wait until the problem is found. The deploy works, it just takes a long time and returns an error. 
Maybe it will become a problem when more and more instances are updated to 2.2.1 and more people have the bug.

Sorry for not opening a new issues and posting it here in the wrong one.

**Comment by sanyafifa** [comment]
@knolleary 
Thanks for responding :)
Deployment took longer. For me it's not critical.
How can you help with debugging?

**Comment by sanyafifa** [comment]
@Steve-Mcl 
Tell the command how to downgrade correctly?

**Comment by Paul-Reed** [comment]
In `node_modules`, I see mqtt.js, but checking it's package.json, it's showing v4.2.8 and not v4.3.5 as mentioned above.

**Comment by sanyafifa** [comment]
@Steve-Mcl 
version 4.2.8
![image](https://user-images.githubusercontent.com/49896945/154248206-7fac9ae1-c7d0-4ba6-b768-10b757212030.png)

**Comment by knolleary** [comment]
The version of the `mqtt` module we care about is under the Node-RED install, *not* in your `.node-red` directory.

Please have a look under `/usr/lib/node_modules/` and/or `/usr/lib/node_modules/node-red/`

**Comment by sanyafifa** [comment]
![image](https://user-images.githubusercontent.com/49896945/154250875-2e7ad5b2-ef60-4d25-8b52-a92303197167.png)

**Comment by Paul-Reed** [comment]
@Steve-Mcl so is that `npm install mqtt@4.3.4 -g` from `/usr/lib/node_modules/node-red` ?

**Comment by Steve-Mcl** [comment]
No...

`npm install mqtt@4.3.4` from /usr/lib/node_modules/node-red

**Comment by Paul-Reed** [comment]
`sudo` is required, otherwise is fails through lack of permission
But downgraded to v4.3.4, restarted and tested.
Can confirm that the Error message still re-occurs.

**Comment by Xploder** [comment]
I'm also encountering this issue since updating to the newest docker container. nodered 2.2.1

**Comment by Steve-Mcl** [comment]
@Paul-Reed did you leave off the ` -g`

**Comment by Paul-Reed** [comment]
Yes.  `sudo npm install mqtt@4.3.4`
I've also double checked the downgraded version number in it's package.json - (4.3.4)

**Comment by sanyafifa** [comment]
i downgraded the mqtt version to 4.3.4. 
`sudo npm install mqtt@4.3.4`
The error is still there :(
![image](https://user-images.githubusercontent.com/49896945/154270391-c1472d99-4820-44da-ad3f-6531179548f0.png)

**Comment by Steve-Mcl** [comment]
An updated for those waiting a solution.

I now understand the issue (race condition between in/out nodes closing vs broker disconnecting AND MQTT lib change of when it raises close events between v4.2 and v4.3) and I am working on a number of angles / solutions. 

TVH: I rarely use full deploy so hadn't noticed this however I suspect the issue has been present since I added Dynamic subscriptions to MQTT 4 months ago 😱 

Should have an update for you today or tomorrow 🤞

**Comment by Paul-Reed** [comment]
Thanks Steve, however I'd be surprised if it has been present 4 months, as I updated to node-RED v2.2.0 on 29th Jan, and this problem did not occur until I updated to v2.2.1 yesterday.
Of late, I've been rebooting my Pi quite a lot, and am confident that I would have noticed.

**Comment by Steve-Mcl** [comment]
Have you recently started using FULL DEPLOY?  Or recently started changing MQTT nodes? 

The issue is related to when either a FULL DEPLOY or a PARTIAL/NODE DEPLOY causes all MQTT in/out nodes to be `close`d 

In other words, if you have been moving flows around and happened to affect all MQTT in/out nodes that are related to 1 broker, you can trigger this timing issue.

No matter anyhow. Perhaps it is better to say, all the elements of the problem have been in place since dynamic MQTT however a number of other factors (MQTT client internal changes) have highlighted the issue.

**Comment by sanyafifa** [comment]
mosqutto log using FULL DEPLOY NR:
`1645025868: Client mqttjs_24e6b775 disconnected.
1645025884: New connection from 127.0.0.1:60960 on port 1883.
1645025884: New connection from 127.0.0.1:60962 on port 1883.
1645025884: New client connected from 127.0.0.1:60960 as mqttjs_ad2fe142 (p2, c1, k60).
1645025884: New client connected from 127.0.0.1:60962 as nodered_897808d04322b0bc (p2, c1, k15).`

> Have you recently started using FULL DEPLOY?

This issue was introduced in v2.2.1 . i always use full deploy

**Comment by Steve-Mcl** [comment]
Update: I have a working fix but I will be spending a lot more time trying various scenarios, including the aedes issue mentioned earlier in this thread before I publish. Possibly tomorrow or Friday. 

In the meantime, please sit tight (or avoid full deploys if you can)

**Comment by goxofy** [comment]
Same problem.  Thanks!

**Comment by denonbw** [comment]
2.2.2 Works great! Thanks for the quick fix

**Comment by quadhammer** [comment]
I'm using the Home Assistant Node Red plugin.  I get the error if I go from nodered_11.0.2 to 11.0.3.

I'm using https://www.home-assistant.io/integrations/mqtt as my broker.

**Comment by Steve-Mcl** [comment]
@quadhammer those version numbers are home assistant versions not node-red versions. What version number of node-red  do you see if you look in the nodes hamburger menu?

**Comment by jonascoenen005** [comment]
when is v2.2.2 being released in home assistant?

**Comment by hardillb** [comment]
@jonascoenen005 that is not a question the Node-RED community can answer, we have no input into when Home Assistant release things

**Comment by quadhammer** [comment]
> @quadhammer those version numbers are home assistant versions not node-red versions. What version number of node-red do you see if you look in the nodes hamburger menu?

Node-RED v2.2.0

**Comment by Steve-Mcl** [comment]
> I get the error if I go from nodered_11.0.2 to 11.0.3.

> Node-RED v2.2.0

@quadhammer  hopefully now you can see you are not on 2.2.2 which contains this fix. The node-red community has no control or influence on when HA release updates. 

In the short term, until HA release an update, I suggest you stick to node deploys (instead of full deploys) to reduce the occurrence of this timeout.

**Comment by quadhammer** [comment]
> now you can see you are not on 2.2.2 which contains this fix.

Yes, thanks.  The latest one is: "Bump node-red from 2.2.0 to 2.2.1 in /node-red" so it shouldn't be too far behind.  I'm sticking with 2.2.0 which doesn't seem to have the problem, even with full deploys.

Cheers!

**Comment by Mark-Muc** [comment]
I do have the same problem 
`Error stopping node: Close timed out`
using:
```
node-red-contrib-tasmota
0.9.12
Tasmota Button
Tasmota Generic
Tasmota Light
Tasmota Sensor
Tasmota Switch
**tasmota-mqtt-broker**
```
some Tasmotadevices won't be found anymore.
2.2.2: Maintenance Release
Nodes
Fix "close timed out" error when performing full deploy or modifying broker node. (#3451) @Steve-Mcl

**Comment by Steve-Mcl** [comment]
@Mark-Muc this is more likely an issue with that node & you should raise it on the tasmota nodes repository.

**Comment by Steve-Mcl** [comment]
Also, @Mark-Muc to minimise / reduce the problem, use "modified nodes" instead of "full deploy" where possible.

![image](https://user-images.githubusercontent.com/44235289/172800173-4514f0e4-c93c-4e77-b33b-e2d59d1b67b5.png)

**Comment by Mark-Muc** [comment]
@Steve-Mcl 
that's what i do normaly . but to geht the devices online i habe to use "Full"
Mark_Muc
